developer/src/kmc-kmn/src/kmw-compiler/validate-layout-file.ts

```diff
@@ -46,6 +46,16 @@ function GetKeyIdUnicodeType(value: string): TKeyIdType {
 
 function KeyIdType(FId: string): TKeyIdType {   // I4142
   FId = FId.toUpperCase();
+
+  // Validate key id format:
+  // K_xxxx -- predefined virtual key - restricted character set
+  // T_xxxx -- custom 'touch' virtual key - touch key id
+  // U_ABCD_1234 -- Unicode key id (1+ chars)
+  // x00 -- "ISO" key identifier (not currently supported in touch layout files)
+  if(!/^((K_[A-Z0-9_?]+)|(T_\S+)|(U_[0-9A-F_]+))$/.test(FId)) {
```

Comment:
Is there an upper limit to the Unicode key ID? U_10FFFF?

Reply:
Yes, that is already in place, see 

https://github.com/keymanapp/keyman/blob/74f4392d752613d45f627c42a1c4eb85ad2f62f6/developer/src/kmc-kmn/src/kmw-compiler/validate-layout-file.ts#L27-L30